docs/blog/posts/03-deploy-llms-on-a-budget.md

```diff
@@ -0,0 +1,319 @@
+
+```bash
+(nos-py38) nos-playground/examples/llm-streaming-chat $ tree .
+├── models
+│   └── phixtral_chat.py
+├── serve.phixtral.yaml
+```
+
+The entire chat interface is defined in the `StreamingChat` module in [`phixtral_chat.py`](https://github.com/autonomi-ai/nos-playground/blob/main/examples/llm-streaming-chat/models/phixtral_chat.py), where the `chat` method returns a string iterable for the gRPC / HTTP server to stream back model predictions to the client. 
+
+The [`serve.phixtral.yaml`](https://github.com/autonomi-ai/nos-playground/blob/main/examples/llm-streaming-chat/serve.phixtral.yaml) serve specification defines the custom chat model, and a custom runtime that NOS uses to execute our model. Follow the annotations below to understand the different components of the serve specification.
+
+<div class="annotate" markdown>
+
+```yaml title="serve.phixtral.yaml"
+
+images: (1)
+  llm-py310-cu121: (2)
+    base: autonomi/nos:latest-py310-cu121 (3)
+    pip: (4)
+      - bitsandbytes
+      - transformers
+      - einops
+      - accelerate
+
+models: (5)
+  mlabonne/phixtral-4x2_8: (6)
+    model_cls: StreamingChat (7)
+    model_path: models/phixtral_chat.py (8)
+    init_kwargs:
+      model_name: mlabonne/phixtral-4x2_8
+    default_method: chat
+    runtime_env: llm-gpu
+    deployment: (9)
+      resources:
+        device: auto
+        device_memory: 7Gi (10)
+```
+</div>
+1. Specifies the custom runtime images that will be used to serve our model.
+2. Specifies the name of the custom runtime image.
+3. Specifies the base image to use for the custom runtime image.
+4. Specifies the pip dependencies to install in the custom runtime image.
+5. Specifies all the custom models we intend to serve.
+6. Specifies the unique name of the custom model (model identifier).
+7. Specifies the model class to use for the custom model.
+8. Specifies the path to the model class definition.
+9. Specifies the deployment resources needed for the custom model.
+10. Specifies the GPU memory to allocate for the custom model.
+
+### 2. Test your custom model locally with NOS
+
+In order to start the NOS server locally, we can simply run the following command:
+
+```bash
+nos serve up -c serve.phixtral.yaml --http
+```
+
+This will build the custom runtime image, and start the NOS server locally, exposing an OpenAI compatible HTTP proxy on port `8000`. This will allow you to chat with your custom LLM endpoint using any OpenAI API compatible client. 
+
+### 3. Deploy your NOS service with SkyPilot
+
+Now that we have defined our serve YAML specification, let's deploy this service on GCP using [SkyPilot](https://github.com/autonomi-ai/nos). In this example, we're going to use SkyPilot's `sky serve` command to deploy our NOS service on spot (pre-emptible) instances on GCP. 
```

Comment:
nit: skypilot link goes to NOS, not skypilot

Reply:
Done.